cartography/intel/rapid7/util.py

```diff
@@ -0,0 +1,292 @@
+"""
+cartography/intel/rapid7/util
+"""
+# pylint: disable=invalid-name,broad-except
+import json
+import logging
+import re
+from typing import Any
+from typing import List
+from typing import Tuple
+
+import pandas
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+def extract_sub_from_resourceid(resourceid: str) -> str:
+    """
+    Get Azure subscription from Azure resource id
+    """
+    if isinstance(resourceid, str):
+        try:
+            match = re.search(
+                "/subscriptions/([0-9a-f-].*?)/resourceGroups/",
+                resourceid,
+                flags=re.IGNORECASE,
+            )
+            if match:
+                sub = match.group(1)
+                return sub
+        except AttributeError as exception:
+            logging.exception("exception: %s", exception)
+    return ""
+
+
+def extract_rg_from_resourceid(resourceid: str) -> str:
+    """
+    Get Azure subscription from Azure resource id
+    """
+    if isinstance(resourceid, str):
+        try:
+            match = re.search(
+                "/resourceGroups/(.*?)/providers/",
+                resourceid,
+                flags=re.IGNORECASE,
+            )
+            if match:
+                rg = match.group(1)
+                return rg
+        except AttributeError as exception:
+            logging.exception("exception: %s", exception)
+    return ""
+
+
+def extract_rapid7_configurations_subscriptionid(configurations: Any[str, list]) -> str:
+    """
+    Extract from json, specific keyvalue pair: azure subscription id
+
+    Input: dataframe row configurations column
+    """
+
+    if configurations is None or not isinstance(configurations, list):
+        return ""
+
+    subscription_id = ""
+    for line in configurations:
+        if line["name"] == "azure":
+            logger.warning("line: %s", line)
+            if isinstance(line["value"], str):
+                azure = json.loads(line["value"])
+            else:
+                azure = line["value"]
+            subscription_id = extract_sub_from_resourceid(azure["resourceId"])
+            logger.warning(
+                "extract_rapid7_configurations_subscriptionid0: %s (%s)",
+                subscription_id,
+                azure["resourceId"],
+            )
+
+    logger.warning("extract_rapid7_configurations_subscriptionid: %s", subscription_id)
+    return subscription_id
+
+
+def extract_rapid7_configurations_resourcegroup(configurations: Any[str, list]) -> str:
+    """
+    Extract from json, specific keyvalue pair: azure resource group
+
+    Input: dataframe row configurations column
+    """
+
+    if configurations is None or not isinstance(configurations, list):
```

Comment:
Does the data returned from rapid7s api sometimes have empty fields?

Reply:
yes

Reply:
fine to resolve?